x-pack/plugins/encrypted_saved_objects/server/create_migration.ts

```diff
@@ -79,21 +81,44 @@ export const getCreateMigration = (
     const decryptDescriptor = { id, type, namespace: decryptDescriptorNamespace };
     const encryptDescriptor = { id, type, namespace: encryptedDoc.namespace };
 
+    let shouldEncrypt = true;
+
     // decrypt the attributes using the input type definition
-    // then migrate the document
-    // then encrypt the attributes using the migration type definition
-    return mapAttributes(
-      migration(
-        mapAttributes(encryptedDoc, (inputAttributes) =>
-          inputService.decryptAttributesSync<any>(decryptDescriptor, inputAttributes, {
+    // if an error occurs during decryption, use the shouldMigrateIfDecryptionFails flag
+    // to determine whether to throw the error or return the un-decrypted document for migration
+    // if we are migrating the un-decrypted document, set `shouldEncrypt` to false to avoid encrypting
+    // an un-decrypted document
+    const documentToMigrate = mapAttributes(encryptedDoc, (inputAttributes) => {
+      try {
+        const decryptedDoc = inputService.decryptAttributesSync<any>(
+          decryptDescriptor,
+          inputAttributes,
+          {
             convertToMultiNamespaceType,
-          })
-        ),
-        context
-      ),
-      (migratedAttributes) =>
-        migratedService.encryptAttributesSync<any>(encryptDescriptor, migratedAttributes)
-    );
+          }
+        );
+        shouldEncrypt = true;
+        return decryptedDoc;
+      } catch (err) {
+        if (!shouldMigrateIfDecryptionFails) {
+          throw err;
+        }
+
+        context.log.warning(
+          `decryption failed for encryptedSavedObject ${encryptedDoc.id} of type ${encryptedDoc.type} with error: ${err.message}. Migration will be applied to the un-decrypted document but this may cause decryption errors later on.`
+        );
+        shouldEncrypt = false;
+        return encryptedDoc.attributes;
+      }
+    });
+
+    // migrate the document
+    // then if `shouldEncrypt` is true, encrypt the attributes using the migration type definition
+    return mapAttributes(migration(documentToMigrate, context), (migratedAttributes) => {
+      return shouldEncrypt
+        ? migratedService.encryptAttributesSync<any>(encryptDescriptor, migratedAttributes)
+        : migratedAttributes;
```

Comment:
@azasypkin We did consider stripping the secrets and essentially disabling `alerts` and `actions`. The concern is that in order to correctly disable alerts, we should also be deleting the scheduled task associated with the rule and invalidating the API key, otherwise we're leaving dangling references. These didn't feel like things we should be doing as part of a migration. WDYT?

Reply:
@azasypkin I added `stripOrDecryptAttributesSync` in [this commit](https://github.com/elastic/kibana/pull/105968/commits/7990aa333ae38dc4019fc9d4d21425d85be80096) and made use of it in [this commit](https://github.com/elastic/kibana/pull/105968/commits/b9b08b838e617bbfe9d28a546f84ef3a0ac3b717)

I originally tried using `stripOrDecryptAttributes` in place of `decryptAttributesSync` when `shouldMigrateIfDecryptionFails = true` but it was not providing the expected behavior. I was expecting it to only strip the `apiKey` attribute from an alert if decryption failed and return a decryption error. Instead it was always stripping the `apiKey` attribute and not even trying to decrypt it. It looks like it is because it's part of the alert SO type's `attributesToStrip` Set. Is this expected?

Instead, I modified the try/catch to still use `decryptAttributesSync`, catch any encryption errors and only then use `stripOrDecryptAttributes`.

Reply:
> I originally tried using stripOrDecryptAttributes in place of decryptAttributesSync when shouldMigrateIfDecryptionFails = true but it was not providing the expected behavior. I was expecting it to only strip the apiKey attribute from an alert if decryption failed and return a decryption error. Instead it was always stripping the apiKey attribute and not even trying to decrypt it. It looks like it is because it's part of the alert SO type's attributesToStrip Set. Is this expected?

Ah, `attributesToStrip`, you're right, completely forgot about that!

> Instead, I modified the try/catch to still use decryptAttributesSync, catch any encryption errors and only then use stripOrDecryptAttributes.

Sounds good, we may want optimize this a bit in the future (to not try to decrypt twice), but Security team can do that in a follow-up if we ever feel the need in this optimization.